core/src/main/scala/akka/persistence/jdbc/journal/dao/DefaultJournalDao.scala

```diff
@@ -48,21 +52,34 @@ class DefaultJournalDao(
   val queries =
     new JournalQueries(profile, journalConfig.eventJournalTableConfiguration, journalConfig.eventTagTableConfiguration)
 
-  override def delete(persistenceId: String, maxSequenceNr: Long): Future[Unit] = {
-    val actions: DBIOAction[Unit, NoStream, Effect.Write with Effect.Read] = for {
-      _ <- queries.markJournalMessagesAsDeleted(persistenceId, maxSequenceNr)
-      highestMarkedSequenceNr <- highestMarkedSequenceNr(persistenceId)
-      _ <- queries.delete(persistenceId, highestMarkedSequenceNr.getOrElse(0L) - 1)
-    } yield ()
+  override def delete(persistenceId: String, toSequenceNr: Long): Future[Unit] = {
+
+    // note: the passed toSequenceNr will be Long.MaxValue when doing a 'full' journal clean-up
+    // see JournalSpec's test: 'not reset highestSequenceNr after journal cleanup'
+    val actions: DBIOAction[Unit, NoStream, Effect.Write with Effect.Read] =
+      highestSequenceNrAction(persistenceId)
+        .flatMap {
+          // are we trying to delete the highest or even higher seqNr ?
+          case highestSeqNr if highestSeqNr <= toSequenceNr =>
+            // if so, we delete up to the before last and
+            // mark the last as logically deleted preserving highestSeqNr
+            queries
+              .delete(persistenceId, highestSeqNr - 1)
+              .flatMap(_ => queries.markAsDeleted(persistenceId, highestSeqNr))
+          case _ =>
+            // if not, we delete up to the requested seqNr
+            queries.delete(persistenceId, toSequenceNr)
+        }
+        .map(_ => ())
```

Comment:
Tests for h2 are passing now locally. Should pass for the the other databases as well. All failures were comping from the akka journal specs.